sdks/java/core/src/main/java/org/apache/beam/sdk/io/TFRecordIO.java

```diff
@@ -243,47 +242,70 @@ public void populateDisplayData(DisplayData.Builder builder) {
 
     @AutoValue.Builder
     abstract static class Builder {
-      abstract Builder setFilenamePrefix(ValueProvider<String> filenamePrefix);
+      abstract Builder setOutputPrefix(ValueProvider<ResourceId> outputPrefix);
+
+      abstract Builder setShardTemplate(String shardTemplate);
 
       abstract Builder setFilenameSuffix(String filenameSuffix);
 
       abstract Builder setNumShards(int numShards);
 
-      abstract Builder setShardTemplate(String shardTemplate);
-
       abstract Builder setCompressionType(CompressionType compressionType);
 
       abstract Write build();
     }
 
     /**
-     * Writes to TFRecord file(s) with the given prefix. This can be a local filename
-     * (if running locally), or a Google Cloud Storage filename of
-     * the form {@code "gs://<bucket>/<filepath>"}
-     * (if running locally or using remote execution).
+     * Writes TFRecord file(s) with the given output prefix. The {@code prefix} will be used as a
+     * to generate a {@link ResourceId} using any supported {@link FileSystem}.
      *
-     * <p>The files written will begin with this prefix, followed by
-     * a shard identifier (see {@link #withNumShards(int)}, and end
-     * in a common extension, if given by {@link #withSuffix(String)}.
+     * <p>In addition to their prefix, created files will have a shard identifier (see
+     * {@link #withNumShards(int)}), and end in a common suffix, if given by
+     * {@link #withSuffix(String)}.
+     *
+     * <p>For more information on filenames, see {@link DefaultFilenamePolicy}.
      */
-    public Write to(String filenamePrefix) {
-      return to(StaticValueProvider.of(filenamePrefix));
+    public Write to(String outputPrefix) {
+      try {
```

Comment:
Same comments as Avro. Also, this looks prettier than the version in Avro. Also, could consider adding a function to FileSystems.